packages/backend-defaults/src/entrypoints/actions/DefaultActionsService.ts

```diff
@@ -0,0 +1,139 @@
+  private constructor(
+    private readonly discovery: DiscoveryService,
+    private readonly config: RootConfigService,
+    private readonly logger: LoggerService,
+    private readonly auth: AuthService,
+  ) {}
+
+  static create({
+    discovery,
+    config,
+    logger,
+    auth,
+  }: {
+    discovery: DiscoveryService;
+    config: RootConfigService;
+    logger: LoggerService;
+    auth: AuthService;
+  }) {
+    return new DefaultActionsService(discovery, config, logger, auth);
+  }
+
+  async list({ credentials }: { credentials: BackstageCredentials }) {
+    const pluginSources =
+      this.config.getOptionalStringArray('backend.actions.pluginSources') ?? [];
+
+    const remoteActionsList = await Promise.all(
+      pluginSources.map(async source => {
+        try {
+          const response = await this.makeRequest({
+            path: `/.backstage/actions/v1/actions`,
+            pluginId: source,
+            credentials,
+          });
+          if (!response.ok) {
+            throw await ResponseError.fromResponse(response);
+          }
+          const { actions } = (await response.json()) as {
+            actions: ActionsServiceAction;
+          };
+
+          return actions;
+        } catch (error) {
+          this.logger.warn(`Failed to fetch actions from ${source}`, error);
+          return [];
+        }
+      }),
+    );
+
+    return { actions: remoteActionsList.flat() };
+  }
+
+  async invoke(opts: {
+    id: string;
+    input?: JsonObject;
+    credentials: BackstageCredentials;
+  }) {
+    const pluginId = this.pluginIdFromActionId(opts.id);
+    const response = await this.makeRequest({
+      path: `/.backstage/actions/v1/actions/${encodeURIComponent(
+        opts.id,
+      )}/invoke`,
+      pluginId,
+      credentials: opts.credentials,
+      options: {
+        method: 'POST',
+        body: JSON.stringify(opts.input),
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      },
+    });
+
+    if (!response.ok) {
+      throw await ResponseError.fromResponse(response);
+    }
+
+    const { output } = await response.json();
+    return { output };
+  }
+
+  private async makeRequest(opts: {
+    path: string;
+    pluginId: string;
+    options?: RequestInit;
+    credentials: BackstageCredentials;
+  }) {
+    const { path, pluginId, credentials, options } = opts;
+    const baseUrl = await this.discovery.getBaseUrl(pluginId);
+
+    const { token } = await this.auth.getPluginRequestToken({
+      onBehalfOf: credentials,
+      targetPluginId: opts.pluginId,
+    });
+
+    return fetch(`${baseUrl}${path}`, {
+      ...options,
+      headers: {
+        ...options?.headers,
+        Authorization: `Bearer ${token}`,
+      },
+    });
+  }
+
+  private pluginIdFromActionId(id: string): string {
+    const colonIndex = id.indexOf(':');
+    if (colonIndex === -1) {
+      throw new Error(`Invalid action id: ${id}`);
```

Comment:
Could this error be more informative? Why is it invalid? What makes it invalid? 

I suspect if I saw this error, I'd have to inspect the source to find out what's wrong with it currently.